lnst/Tests/PktGen.py

```diff
@@ -0,0 +1,157 @@
+import re
+import time
+import logging
+from typing import Iterator
+from subprocess import Popen
+
+from lnst.Common.Utils import kmod_in_use
+from lnst.Common.IpAddress import ipaddress, Ip4Address
+from lnst.Tests.BaseTestModule import BaseTestModule, TestModuleError
+from lnst.Common.Parameters import IntParam, IpParam, StrParam, ListParam
+
+
+class PktGenResultsParser:
+    def __init__(self, devs: list[str], res: dict) -> None:
+        self._devs = devs
+        self._res = res
+
+    def _read_dev_outputs(self) -> Iterator[tuple[str, str]]:
+        for device in self._devs:
+            output = ""
+            with open(f"/proc/net/pktgen/{device}", "r") as f:
+                output = f.readlines()
+            yield device, "\n".join(output)
+
+    def _split_output(self, output: str) -> tuple:
+        match = re.search(r"Params:(.+)Current:(.+)Result:\s(\w+)", output, re.DOTALL)
+        if not match:
+            raise TestModuleError("Could not parse pktgen devide output")
+
+        return match.groups()
+
+    def _parse_values(self, params) -> dict[str, str]:
+        values = {}
+
+        for key, value in re.findall(r"(\w+):?\s(\S+)", params, re.MULTILINE):
+            values[key.lower()] = value
+
+        return values
+
+    def parse_dev_outputs(self):
+        for device, output in self._read_dev_outputs():
+            params, current, state = self._split_output(output)
+
+            self._res[device] = {
+                "params": self._parse_values(params),
+                "current": self._parse_values(current),
+                "state": state,
+            }
+
+
+class PktGen(BaseTestModule):
+    """
+    In the scope of this module, the physical interface is refered as `interface`.
+    Pktgen device (interface@anything) is refered as device.
+
+    Inspired by https://github.com/torvalds/linux/blob/master/samples/pktgen/pktgen_sample03_burst_single_flow.sh
+    """
+
+    cpus = ListParam(type=IntParam())  # each CPU is 1 generator
+
+    pgctrl = StrParam(mandatory=False, default="/proc/net/pktgen/pgctrl")
+
+    src_mac = StrParam(mandatory=False, default="90:e2:ba:ff:ff:ff")
+    dst_mac = StrParam()
+
+    src_ip = IpParam(
+        mandatory=False, default=ipaddress("192.168.1.1")
+    )  # IP doesn't really matter
+    dst_ip = IpParam()
+
+    interface = StrParam()
```

Comment:
this could be a `DeviceParam` which would give you access to both the nic name as well as the mac address and the ip address.

Reply:
Nice catch, thanks.